File: src/pricepoint/api/schemas/saved.py
"""Pydantic models for saved-property endpoints."""

from datetime import datetime

from pydantic import BaseModel


class SavedPropertyCreate(BaseModel):
    """Request body to save/bookmark a property."""

    listing_id: int
    notes: str | None = None


class SavedPropertyUpdate(BaseModel):
    """Request body to update notes on a saved property."""

    notes: str | None = None


class SavedPropertyResponse(BaseModel):
    """Single saved-property record returned to the client."""

    id: int
    listing_id: int
    notes: str | None = None
    created_at: datetime
    listing_address: str | None = None

    # Enriched fields from RedfinListing
    city: str | None = None
    state: str | None = None
    zip_code: str | None = None
    listing_status: str | None = None
    listing_price: float | None = None
    sold_price: float | None = None
    num_beds: int | None = None
    num_baths: float | None = None
    sqft: int | None = None
    year_built: int | None = None
    photo_url: str | None = None
    lat: float | None = None
    lon: float | None = None

    model_config = {"from_attributes": True}

File: frontend/src/services/saved.ts
import axios from "axios";

const client = axios.create({
  baseURL: "/",
  headers: { "Content-Type": "application/json" },
});

export interface SavedPropertyResponse {
  id: number;
  listing_id: number;
  notes: string | null;
  created_at: string;
  listing_address: string | null;
  city: string | null;
  state: string | null;
  zip_code: string | null;
  listing_status: string | null;
  listing_price: number | null;
  sold_price: number | null;
  num_beds: number | null;
  num_baths: number | null;
  sqft: number | null;
  year_built: number | null;
  photo_url: string | null;
  lat: number | null;
  lon: number | null;
}

function authHeaders(token: string) {
  return { Authorization: `Bearer ${token}` };
}

export async function getSavedProperties(token: string): Promise<SavedPropertyResponse[]> {
  const { data } = await client.get<SavedPropertyResponse[]>("/api/saved", {
    headers: authHeaders(token),
  });
  return data;
}

export async function saveProperty(
  token: string,
  listingId: number,
  notes?: string,
): Promise<SavedPropertyResponse> {
  const { data } = await client.post<SavedPropertyResponse>(
    "/api/saved",
    { listing_id: listingId, notes: notes ?? null },
    { headers: authHeaders(token) },
  );
  return data;
}

export async function deleteSavedProperty(token: string, savedId: number): Promise<void> {
  await client.delete(`/api/saved/${savedId}`, {
    headers: authHeaders(token),
  });
}
